### ISSUE TITLE
Inconsistent Table Alias Usage with `order`

### ISSUE THREAD

**Issue by jherdman** [rails/rails#53509]
### Steps to reproduce

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails"
  # If you want to test against edge Rails replace the previous line with this:
  # gem "rails", github: "rails/rails", branch: "main"

  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.string :title
  end

  create_table :comments, force: true do |t|
    t.string :title
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class BugTest < ActiveSupport::TestCase
  def test_association_stuff
    post = Post.create!(title: "Hello")
    post.comments << (comment_1 = Comment.create!(title: "Hello, comment"))

    post_2 = Post.create!(title: "World")
    post_2.comments << (comment_2 = Comment.create!(title: "World, comment"))

    # WORKS
    assert_equal [comment_2, comment_1], Comment.joins(:post).where.not(post: { title: nil }).order('post.title DESC')

    # WORKS, but...
    assert_equal [comment_2, comment_1], Comment.joins(:post).order('posts.title DESC')

    # BROKEN... `post` alias isn't set
    assert_equal [comment_2, comment_1], Comment.joins(:post).order('post.title DESC')
  end
end

```

### Expected behavior

Table aliasing should be consistent between `order` and `where`.

### Actual behavior

It's not 😆 . What we see is that the `order` case _didn't_ use a table alias. This makes constructing filtering and sorting queries difficult, especially when they're created by the user (think a table with filters and optional sorting).

An example:

```ruby
class Comment < ApplicationRecord
  scope :order_by_post_title, ->(direction) { joins(:post).order('post.title') }
  scope :with_titled_post, -> { joins(:post).where.not(post: { title: nil }) }
end
```

In this case `order_by_post_title` only works when `with_titled_post` is used. The developer might then try...

```ruby
class Comment < ApplicationRecord
  scope :order_by_post_title, ->(direction) { joins(:post).order('posts.title') }
  scope :with_titled_post, -> { joins(:post).where.not(post: { title: nil }) }
end
```

And now the opposite is true!

The ideal scenario is that I don't really need to know what ActiveRecord is doing under the hood. The ideal is:

```ruby
class Comment < ApplicationRecord
  scope :order_by_post_title, ->(direction) { joins(:post).order(post: { title: direction }) } # Hash, just like where
  scope :with_titled_post, -> { joins(:post).where.not(post: { title: nil }) }
end
```

### System configuration

**Rails version**: 7.1+ (I've tested on v8 too)

**Ruby version**: 3.3.4

**Comment by ibrahima** [comment]
Hi! I've just noticed a similar issue with `select` - if I have a table alias and reference the aliased table in a `where` call, the alias is applied, but if not, then it uses the original table name. This makes conditionally chaining queries annoying depending on what conditions are applied. E.g. in my application, I have different filters applied to a query depending on the user, and then I `select` some additional fields onto the object from the joined relations, but I have to reference the joined table with different names depending on whether the table was referenced in the `where` clauses.

In this test case, I have two models, Post, and PostComment, but the relation on Post is just called `comments`. If I add the `where` clause then I can include the joined table fields using the `comments` alias, but otherwise I have to use the original `post_comments` table name.

```
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails"
  # If you want to test against edge Rails replace the previous line with this:
  # gem "rails", github: "rails/rails", branch: "main"

  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.string :title
  end

  create_table :post_comments, force: true do |t|
    t.string :title
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  has_many :comments, class_name: "PostComment"
end

class PostComment < ActiveRecord::Base
  belongs_to :post
end

class BugTest < ActiveSupport::TestCase
  def test_association_stuff
    post = Post.create!(title: "Hello")
    post.comments << (comment_1 = PostComment.create!(title: "Hello, comment"))

    post_2 = Post.create!(title: "World")
    post_2.comments << (comment_2 = PostComment.create!(title: "World, comment"))

    # WORKS
    assert_equal [post, post_2], Post.joins(:comments).where.not(comments: { title: nil }).select(posts: ["*"], comments: ["title as comment"])

    # BROKEN... `comments` alias isn't set
    assert_equal [post, post_2], Post.joins(:comments).select(posts: ["*"], comments: ["title as comment"])
  end
end
```

**Comment by ibrahima** [comment]
@kamipo by the way, it seems like the latest changes on `main` have broken my similar test case further - now it turns `select(posts: ["*"])` into `SELECT "posts"."*"` instead of `SELECT posts.*` which fails because `"*"` is not a column name. Now, I don't know if it was ever intended for `select` to work that way, I was kind of pleasantly surprised when I wrote that query and it worked, but just wanted to mention it in case that's something that should still be working. Thank you!

**Comment by kamipo** [comment]
@ibrahima Thanks for the report, that issue will be fixed by #53526.

**Comment by jherdman** [comment]
Will this be released as a bug fix at some point?

**Comment by kamipo** [comment]
@jherdman This has been released in 8.0.
